data/model/autoprune.py

```diff
@@ -218,3 +232,199 @@ def namespace_has_autoprune_task(namespace_id):
 
 def create_autoprune_task(namespace_id):
     AutoPruneTaskStatus.create(namespace=namespace_id, status="queued", last_ran_ms=None)
+
+
+def update_autoprune_task(task, task_status):
+    try:
+        task.status = task_status
+        task.last_ran_ms = get_epoch_timestamp_ms()
+        task.save()
+    except AutoPruneTaskStatus.DoesNotExist:
+        return None
+    except Exception as err:
+        raise Exception(
+            f"Error updating autoprune task for namespace id: {task.namespace_id}, task_status: {task_status} with error as: {str(err)}"
+        )
+
+
+def fetch_autoprune_task(task_run_interval_ms=60 * 60 * 1000):
+    """
+    Get the auto prune task prioritized by last_ran_ms = None followed by asc order of last_ran_ms.
+    task_run_interval_ms specifies how long a task must wait before being ran again.
+    """
+    try:
+        query = (
+            AutoPruneTaskStatus.select(AutoPruneTaskStatus)
+            .join(User, on=(AutoPruneTaskStatus.namespace == User.id))
+            .where(
+                User.enabled == True,
+                (AutoPruneTaskStatus.last_ran_ms < get_epoch_timestamp_ms() - task_run_interval_ms)
+                | (AutoPruneTaskStatus.last_ran_ms.is_null(True)),
+            )
+            .order_by(AutoPruneTaskStatus.last_ran_ms.asc(nulls="first"))
+        )
+        return db_for_update(query, skip_locked=SKIP_LOCKED).get()
+    except AutoPruneTaskStatus.DoesNotExist:
+        return []
+
+
+def fetch_autoprune_task_by_namespace_id(namespace_id):
+    try:
+        return (
+            AutoPruneTaskStatus.select().where(AutoPruneTaskStatus.namespace == namespace_id).get()
+        )
+    except AutoPruneTaskStatus.DoesNotExist:
+        return None
+
+
+def delete_autoprune_task(task):
+    try:
+        task.delete_instance()
+    except AutoPruneTaskStatus.DoesNotExist:
+        return None
+    except Exception as err:
+        raise Exception(
+            f"Error deleting autoprune task for namespace id: {task.namespace_id} with error as: {str(err)}"
+        )
+
+
+def prune_repo_by_number_of_tags(repo, policy_config, namespace, tag_page_limit):
+    policy_method = policy_config.get("method", None)
+
+    if policy_method != AutoPruneMethod.NUMBER_OF_TAGS.value:
+        raise KeyError("Unsupported policy config provided", policy_config)
```

Comment:
[KeyError](https://docs.python.org/3/library/exceptions.html#KeyError) is raised when a mapping (dictionary) key is not found in the set of existing keys.

This exception class does not apply here. Maybe ValueError?

Reply:
Agreed, added `InvalidNamespaceAutoPruneMethod` exception.